File: src/lowlevel/server/icefsoperators/icefsMkDir.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 07:16:00
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsMkDir.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsMkDir(parentFakeInode uint64, name string, mode uint32, fuseEntryParamBuilder FuseEntryParamBuilder) (status int32, entry any) {
	var err error
	s.inodeCacheLock.RLock()
	parentInode := s.getIcefsInode(parentFakeInode)
	if parentInode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}
	parentInode.inodeLock.RLock()
	s.inodeCacheLock.RUnlock()
	err = syscall.Mkdirat(parentInode.fd, name, mode)
	parentInode.inodeLock.RUnlock()
	if err != nil {
		status = icefserror.IcefsStdErrno(err)
		return
	}

	entry, err = s.doIcefsLookUp(parentFakeInode, name, fuseEntryParamBuilder)
	if err != nil {
		status = icefserror.IcefsStdErrno(err)
		return
	}

	status = icefserror.ICEFS_EOK
	return
}

func (s *IcefsGRpcServer) DoIcefsMkDir(ctx context.Context, req *pb.IcefsMkDirReq) (*pb.IcefsMkDirRes, error) {
	var res pb.IcefsMkDirRes
	var entry any

	res.Status, entry = s.server.doIcefsMkDir(req.ParentInode, req.Name, req.Mode, GRpcFuseEntryParamBuilder)
	if res.Status == icefserror.ICEFS_EOK {
		res.Entry = entry.(*pb.FuseEntryParam)
	}

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsMkDir(ctx context.Context, req *icefsthrift.IcefsMkDirReq) (*icefsthrift.IcefsMkDirRes, error) {
	var res icefsthrift.IcefsMkDirRes
	var entry any

	res.Status, entry = s.server.doIcefsMkDir(uint64(req.ParentInode), req.Name, uint32(req.Mode), ThriftFuseEntryParamBuilder)
	if res.Status == icefserror.ICEFS_EOK {
		res.Entry = entry.(*icefsthrift.FuseEntryParam)
	}

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsMkDir.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:40:53
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsMkDir.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsMkDir(fuse_req_t fuseReq, fuse_ino_t parentInode,
                               const char *name, mode_t mode) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsMkDirReq req;
      icefsgrpc::IcefsMkDirRes res;
      grpc::ClientContext ctx;
      struct fuse_entry_param entry;

      req.set_parent_inode(parentInode);
      req.set_name(name);
      req.set_mode(mode);

      grpc::Status status = gRpcClient->DoIcefsMkDir(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        IcefsGRpcFillFuseEntryParamIn(&entry, res.entry());
        fuse_reply_entry(fuseReq, &entry);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsMkDirReq req;
      icefsthrift::IcefsMkDirRes res;
      struct fuse_entry_param entry;

      req.__set_parent_inode(parentInode);
      req.__set_name(name);
      req.__set_mode(mode);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsMkDir(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      if (!res.status) {
        IcefsThriftFillFuseEntryParamIn(&entry, res.entry);
        fuse_reply_entry(fuseReq, &entry);
      } else {
        fuse_reply_err(fuseReq, res.status);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}